### PR TITLE
[SYCL][CUDA] Add tests for asynchronous barrier

### PR DESCRIPTION
**Pull request by t4c1**
Adds tests for https://github.com/intel/llvm/pull/5303

### CODE REVIEW

File: SYCL/GroupAlgorithm/barrier.cpp [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  if (!q.get_device().has(aspect::ext_oneapi_cuda_async_barrier)) {
+    std::cout << "Barrier is not supported by the device. Skipping test."
+              << std::endl;
+    return 0;
+  }
+  basic();
+  interface();
+
+  return 0;
+}
```
Comment (steffenlarsen) [inline SYCL/GroupAlgorithm/barrier.cpp]:
```suggestion
}

```

File: SYCL/GroupAlgorithm/barrier.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    q.submit([&](handler &cgh) {
+      auto acc = buf.get_access<access::mode::read_write>(cgh);
+      accessor<int, 1, access::mode::read_write, access::target::local> loc(
+          N, cgh);
+      accessor<barrier, 1, access::mode::read_write, access::target::local>
+          loc_barrier(2, cgh);
+      cgh.parallel_for(nd_range<1>(N, N), [=](nd_item<1> item) {
+        size_t idx = item.get_local_linear_id();
+        loc[idx] = acc[idx];
+        if (idx <= 2) {
```
Comment (steffenlarsen) [inline SYCL/GroupAlgorithm/barrier.cpp]:
```suggestion
        if (idx < 2) {
```

File: SYCL/GroupAlgorithm/barrier.cpp [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+    q.submit([&](handler &cgh) {
+      auto data_acc = data_buf.get_access<access::mode::read_write>(cgh);
+      auto test1_acc = test1_buf.get_access<access::mode::read_write>(cgh);
+      auto test2_acc = test2_buf.get_access<access::mode::read_write>(cgh);
+      accessor<int, 1, access::mode::read_write, access::target::local> loc(
+          N, cgh);
+      accessor<barrier, 1, access::mode::read_write, access::target::local>
+          loc_barrier(2, cgh);
+      stream s(1024 * 1024, 1024, cgh);
```
Comment (steffenlarsen) [inline SYCL/GroupAlgorithm/barrier.cpp]: Is this used anywhere?

**Review by steffenlarsen — COMMENTED** [review]
The tests look good. I've added some small comments.

One concern is that the the `interface` test case includes a lot in a single kernel. If that test fails it may be hard to determine which part of the test caused the failure. Would it make sense to either split it up or save and validate intermediate results to help isolate the failing operations?

**Comment by t4c1** [comment]
I can split the test into multiple kernels, but a large amount of code (more or less the whole basic test) would need to be duplicated in each test.

**Comment by steffenlarsen** [comment]
> I can split the test into multiple kernels, but a large amount of code (more or less the whole basic test) would need to be duplicated in each test.

Is there no good way of reducing common code for the cases? If it isn't worth it, I am okay with keeping the test as-is.

**Comment by t4c1** [comment]
I think at best I can reduce the amount of code duplication (but not eliminate it) at cost of making the code a bit harder to read. I would prefer to leave this as is, as it is unlikely that implementation of something changes in a way that would break semantics but not interface.

File: SYCL/GroupAlgorithm/barrier.cpp (hunk 3 above)
Comment (t4c1) [reply SYCL/GroupAlgorithm/barrier.cpp]: No, this was left in from debugging. Removed.

**Review by steffenlarsen — APPROVED** [review]

**Pull request merged**